File: app/events/chatting.py
from flask import session
from flask_socketio import send, emit, join_room, leave_room
from datetime import datetime

from app import app, socketio, mongo
from app.routes.message import *


def get_room_name(data):
    url =  data['url']
    base = data['base'] + '/chat/room/'
    room = url.replace(base, '')
    
    return room


@socketio.on('joined', namespace='/chat')
def joined(data):
    room_name = get_room_name(data)
    
    join_room(room_name)

    nickName = session['nickname']
    print(room_name + ', ' + nickName + ',+ 입장했습니다.')
    emit('status', {'room': room_name, 'msg': nickName + '님이 입장했습니다.'}, room=room_name)

    mongo.db.messages.update_many(
        {'room_name': room_name,'receivers.id': session['id']},
        {'$set': {'receivers.$.view': True}}
    )
    mongo.db.rooms.update_many(
        {'name': room_name, 'messages.receivers.id': session['id']},
        {'$set': {'messages.$[].receivers.$.view': True}}        
    )


@socketio.on('left', namespace='/chat')
def left(data):
    nickName = session['nickname']
    room = get_room_name(data)

    print(room + ', ' + nickName + '- 퇴장했습니다.')
    emit('status', {'room': room, 'msg': str(nickName) + '님이 나갔습니다.'}, room=room)
    # db에 이름, 시간 저장


@socketio.on('text', namespace='/chat')
def text(message):
    if 'msg' in message:
        msg = message['msg']
        if msg != '\n' and msg != '':
            nickName = session['nickname']
            room_name = message['room']
            msg = msg.replace('\n', '')
            
            print(room_name + ', ' + nickName + ', ' + msg)
            emit('message', {
                'name': nickName, 
                'msg': msg,
                'profile_image': session['profile_image']
                }, room=room_name)

            room = mongo.db.rooms.find_one({'name': room_name})

            users = list()

            for user in room['users']:
                if user['id'] != session['id']:
                    users.append({
                        'id': user['id'], 
                        'nickname': user['nickname'],
                        'view': False
                    })

            data = {
                'content': msg,
                'room_name': room_name,
                'senders': {
                    'id': session['id'], 
                    'nickname': session['nickname']
                },
                'receivers': users,
                'created_at': datetime.now().strftime('%Y-%m-%d %H:%M:%S')
            }

            mongo.db.messages.insert(data)
            mee = mongo.db.messages.find_one(data)
            mongo.db.rooms.update_one(
                {'name' : room['name']}, 
                {'$push': {'messages': mee}}
            )
            
            if room['group'] == False:
                send_me(msg, room_name)
            else:
                send_friend(room['users'], msg, room_name)

File: app/static/script/chatting.js
var socket;
var room_name;

$(document).ready(function(){   
    var $comments = $('.comments');
    socket = io.connect('http://' + document.domain + ':' + location.port + '/chat');
    
    socket.on('connect', function() {
        socket.emit('joined', {
            url: document.URL, 
            base: 'http://' + document.domain + ':' + location.port, 
            time: new Date()
        });
    });

    socket.on('status', function(data) {
        var message = data.msg;
        room_name = data.room;

        var $div = $(`<div class="ui log label"></div>`)
            .text(message)
        
        $comments.append($div);
        $('.scrolling').scrollTop($('.scrolling')[0].scrollHeight);
    });

    socket.on('message', function(data) {
        var nickName = $("#nickName").text();
        nickName = nickName.replace(/ /gi, '');
        nickName = nickName.replace(/\n/gi, '');

        var name = data.name;
        var profile_image = data.profile_image;
        var message = data.msg;

        var Now = new Date();
        var date = Now.getHours() + ':' + Now.getMinutes();

        var $name = $('<a class="author">').append(name);
        var $date = $('<span class="date">').append(date);
        var $metadata = $('<div class="metadata" style="margin-right: 10px;">').append($date);
        var $name_date = $('<div>').append($name, $metadata);
        
        if (nickName != name) {
            var $text = $('<div class="ui left pointing label text">')
                .append(message);
        
            var $content = $('<div class="content">')
                .append($name_date, $text);

            if (profile_image != '') {
                var $image = $(`<a class="avatar">`).append(
                    `<img class="ui circular image" src=` + profile_image + `>`
                );
            }
            else {
                var $image = $(`<a class="avatar">`).append(
                    `<img class="ui circular image" src=` + '../../static/images/default.png' + `>`
                );
            }

            var $comment = $(`
                <div class="comment">
            `).append($image, $content);
        }
        else {
            var $text = $('<div class="ui right pointing label text" style="background-color: #ffee52">')
                .append(message);
    
            var $content = $('<div class="content">')
                .append($metadata, $text);

            var $comment = $(`
                <div class="comment" style="text-align: right; margin-right: 20px;">
            `).append($content);
        }

        $comments.append($comment);
        $('.scrolling').scrollTop($('.scrolling')[0].scrollHeight);
    });

    $('#submit').click(function() {
        text = $('#text').val();
        $('#text').val('');
        socket.emit('text', {msg: text, room: room_name});
        $('.scrolling').scrollTop($('.scrolling')[0].scrollHeight);
    });

    $('#text').keypress(function(e) {
        var code = e.keyCode || e.which;
        if (code == 13) {
            text = $('#text').val();
            $('#text').val('');
            socket.emit('text', {msg: text, room: room_name});
            $('.scrolling').scrollTop($('.scrolling')[0].scrollHeight);
        }
    });

    $('.scrolling').scrollTop($('.scrolling')[0].scrollHeight);
});

function leave_room() {
    socket.emit('left', {
        url: document.URL, 
        base: 'http://' + document.domain + ':' + location.port
    }, 
    function() {
        socket.disconnect();
        window.location.href = "/";
    });
}